pkg/store/copr/coprocessor.go

```diff
@@ -1113,20 +1146,65 @@ func (it *copIterator) Next(ctx context.Context) (kv.ResultSubset, error) {
 	return resp, nil
 }
 
+func (it *copIterator) liteSendReq(ctx context.Context) (resp *copResponse) {
+	defer func() {
+		r := recover()
+		if r != nil {
+			logutil.Logger(ctx).Error("copIteratorWork meet panic",
+				zap.Any("r", r),
+				zap.Stack("stack trace"))
+			resp = &copResponse{err: util2.GetRecoverError(r)}
+		}
+	}()
+
+	worker := it.liteWorker.worker
+	backoffermap := make(map[uint64]*Backoffer)
+	for {
+		for len(it.respChan) > 0 {
+			select {
+			case resp = <-it.respChan:
+				return resp
+			default:
+			}
+		}
+		if len(it.tasks) == 0 {
+			return nil
+		}
+		curTask := it.tasks[0]
+		if cap(it.respChan) < (len(curTask.batchTaskList) + 2) {
+			it.respChan = make(chan *copResponse, 2+len(curTask.batchTaskList))
```

Comment:
Here we assume that all copResponses are fill to the respChan after calling handleTaskOnce, how about just let handleTaskOnce return a slice of copResponses (instead of using a channel)?

```go
		for len(it.respChan) > 0 {
			select {
			case resp = <-it.respChan:
				return resp
			default:
			}
		}
```

The above code also assume that no copResponse will be pushed into the respChan asynchronously, but actually current code cannot guarantee that (eg. someone may change the impl of handleTaskOnce and let it write to respCh asynchronously).

Reply:
Nice catch, I'm working on this.